File: app/src/main/java/nl/koenhabets/yahtzeescore/activities/StatsActivity.java
package nl.koenhabets.yahtzeescore.activities;

import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.Context;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.TextView;

import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.google.android.material.appbar.AppBarLayout;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.matomo.sdk.Tracker;
import org.matomo.sdk.extra.TrackHelper;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import nl.koenhabets.yahtzeescore.DataManager;
import nl.koenhabets.yahtzeescore.R;
import nl.koenhabets.yahtzeescore.ScoreItem;

public class StatsActivity extends AppCompatActivity {

    private EditText editText1;
    private EditText editText2;
    private EditText editText3;
    private EditText editText4;
    private EditText editText5;
    private EditText editText6;
    private EditText editText21;
    private EditText editText22;
    private EditText editText23;
    private EditText editText24;
    private EditText editText25;
    private EditText editText26;
    private EditText editText27;
    private EditText editText28;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_stats);
        setTitle(getString(R.string.stats));
        try {
            Tracker tracker = MainActivity.getTracker2();
            TrackHelper.track().screen("/stats").title("Stats").with(tracker);
        } catch (Exception e) {
            e.printStackTrace();
        }
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        SharedPreferences sharedPref = getSharedPreferences("nl.koenhabets.yahtzeescore", Context.MODE_PRIVATE);

        JSONArray jsonArray = new JSONArray();
        try {
            jsonArray = new JSONArray(sharedPref.getString("scoresSaved", ""));
        } catch (JSONException e) {
            e.printStackTrace();
        }
        editText1 = findViewById(R.id.editText);
        editText2 = findViewById(R.id.editText3);
        editText3 = findViewById(R.id.editText4);
        editText4 = findViewById(R.id.editText5);
        editText5 = findViewById(R.id.editText6);
        editText6 = findViewById(R.id.editText7);

        editText21 = findViewById(R.id.editText9);
        editText22 = findViewById(R.id.editText10);
        editText23 = findViewById(R.id.editText8);
        editText24 = findViewById(R.id.editText11);
        editText25 = findViewById(R.id.editText12);
        editText26 = findViewById(R.id.editText13);
        editText27 = findViewById(R.id.editText14);
        editText28 = findViewById(R.id.editText16);
        TextView tVYBonus = findViewById(R.id.textView7);
        TextView textViewGraph = findViewById(R.id.textView8);

        LineChart lineChart = findViewById(R.id.chart);
        List<Entry> entries = new ArrayList<>();

        AppBarLayout appBarLayout = findViewById(R.id.appBarLayout);
        appBarLayout.setVisibility(View.GONE);

        if(!sharedPref.getBoolean("yahtzeeBonus", false)){
            tVYBonus.setVisibility(View.GONE);
            editText28.setVisibility(View.GONE);
        }

        JSONObject jsonObject = processScores(jsonArray);
        readScores(jsonObject);

        List<ScoreItem> scoreItemsDate = DataManager.loadScores(this);

        Collections.sort(scoreItemsDate, (o1, o2) -> o1.getDate().compareTo(o2.getDate()));
        float sum = 0;
        int gamesHidden;
        if (scoreItemsDate.size() > 200) {
            gamesHidden = 30;
            textViewGraph.setText(getString(R.string.games_hidden, gamesHidden));
        } else if (scoreItemsDate.size() > 100) {
            gamesHidden = 10;
            textViewGraph.setText(getString(R.string.games_hidden, gamesHidden));
        } else if (scoreItemsDate.size() > 50) {
            gamesHidden = 5;
            textViewGraph.setText(getString(R.string.games_hidden, gamesHidden));
        } else {
            gamesHidden = 0;
        }
        for (int d = 0; d < scoreItemsDate.size(); d++) {
            sum = sum + scoreItemsDate.get(d).getScore();
            float value = sum / (d + 1);
            if (d > gamesHidden - 1) {
                entries.add(new Entry(d, value));
            }

        }
        Log.i("entries", entries.size() + "size");
        LineDataSet dataSet = new LineDataSet(entries, getString(R.string.average_score));
        dataSet.setColor(Color.BLUE);
        dataSet.setValueTextColor(Color.YELLOW);
        LineData lineData = new LineData(dataSet);
        lineChart.setData(lineData);
        lineChart.getDescription().setText(getString(R.string.average_score_of_last, scoreItemsDate.size()));
        int nightModeFlags =
                this.getResources().getConfiguration().uiMode &
                        Configuration.UI_MODE_NIGHT_MASK;
        if (nightModeFlags == Configuration.UI_MODE_NIGHT_YES) {
            lineChart.getXAxis().setTextColor(Color.WHITE);
            lineChart.getAxisLeft().setTextColor(Color.WHITE);
            lineChart.getLegend().setTextColor(Color.WHITE);
            lineChart.getDescription().setTextColor(Color.WHITE);
        }

        lineChart.invalidate();

        disableEdit();
        if (sharedPref.getBoolean("statsInfoDialog", true)) {
            infoDialog(false);
        }
    }

    private void infoDialog(boolean checkboxChecked) {
        LayoutInflater inflater = this.getLayoutInflater();
        AlertDialog.Builder builder = new AlertDialog.Builder(this);

        View view = inflater.inflate(R.layout.dailog_stats, null);
        CheckBox checkBox = view.findViewById(R.id.checkBox);
        checkBox.setChecked(checkboxChecked);
        builder.setView(view);
        builder.setPositiveButton("Ok", (dialog, id) -> {
        });
        checkBox.setOnClickListener(view2 -> {
            SharedPreferences sharedPref = getSharedPreferences("nl.koenhabets.yahtzeescore", Context.MODE_PRIVATE);
            sharedPref.edit().putBoolean("statsInfoDialog", !checkBox.isChecked()).apply();
        });
        builder.show();
    }

    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.help) {
            SharedPreferences sharedPref = getSharedPreferences("nl.koenhabets.yahtzeescore", Context.MODE_PRIVATE);
            infoDialog(!sharedPref.getBoolean("statsInfoDialog", true));
        } else {
            finish();
        }
        return true;
    }

    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_stats, menu);
        return true;
    }

    private JSONObject processScores(JSONArray jsonArray) {
        JSONObject jsonObject = new JSONObject();
        for (int k = 1; k < 7; k++) {
            try {
                jsonObject.put(k + "", proccessField(jsonArray, k));
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        for (int k = 21; k < 29; k++) {
            try {
                jsonObject.put(k + "", proccessField(jsonArray, k));
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        Log.i("atadf", jsonObject.toString());
        return jsonObject;
    }

    private String proccessField(JSONArray jsonArray, int d) {
        double totalScore = 0;
        double scoreCount = 0;
        double scoreCountMax = 0;
        for (int i = 0; i < jsonArray.length(); i++) {

            try {
                JSONObject jsonObject = jsonArray.getJSONObject(i).getJSONObject("allScores");
                String val = jsonObject.getString(d + "");
                if (!val.equals("") && !val.equals("0")) {
                    int valInt = Integer.parseInt(val);
                    totalScore = totalScore + valInt;
                    scoreCount = scoreCount + 1;
                }
                scoreCountMax = scoreCountMax + 1;
            } catch (JSONException ignored) {
            }

        }
        int chance = (int) Math.round((scoreCount / scoreCountMax) * 100.0);
        double average = (totalScore / scoreCountMax);
        return round(average, 1) + "(" + chance + ")";
    }

    public static double round(double value, int places) {
        double res = 0;
        if (value != 0.0 && !Double.isNaN(value)) {
            if (places < 0) throw new IllegalArgumentException();

            BigDecimal bd = new BigDecimal(value);
            bd = bd.setScale(places, RoundingMode.HALF_UP);
            res = bd.doubleValue();
        }
        return res;
    }

    private void readScores(JSONObject jsonObject) {
        Log.i("score", "read" + jsonObject.toString());
        try {
            editText1.setText(jsonObject.getString("1"));
            editText2.setText(jsonObject.getString("2"));
            editText3.setText(jsonObject.getString("3"));
            editText4.setText(jsonObject.getString("4"));
            editText5.setText(jsonObject.getString("5"));
            editText6.setText(jsonObject.getString("6"));
            editText21.setText(jsonObject.getString("21"));
            editText22.setText(jsonObject.getString("22"));
            editText23.setText(jsonObject.getString("23"));
            editText24.setText(jsonObject.getString("24"));
            editText25.setText(jsonObject.getString("25"));
            editText26.setText(jsonObject.getString("26"));
            editText27.setText(jsonObject.getString("27"));
            editText28.setText(jsonObject.getString("28"));
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }

    private void disableEdit() {
        editText1.setEnabled(false);
        editText2.setEnabled(false);
        editText3.setEnabled(false);
        editText4.setEnabled(false);
        editText5.setEnabled(false);
        editText6.setEnabled(false);
        editText21.setEnabled(false);
        editText22.setEnabled(false);
        editText23.setEnabled(false);
        editText24.setEnabled(false);
        editText25.setEnabled(false);
        editText26.setEnabled(false);
        editText27.setEnabled(false);
        editText28.setEnabled(false);

    }
}

File: app/src/main/java/nl/koenhabets/yahtzeescore/activities/StatsActivity.kt
package nl.koenhabets.yahtzeescore.activities

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.res.Configuration
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.CheckBox
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintSet
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import nl.koenhabets.yahtzeescore.MovingAverage
import nl.koenhabets.yahtzeescore.R
import nl.koenhabets.yahtzeescore.ScoreComparatorDate
import nl.koenhabets.yahtzeescore.ScoreItem
import nl.koenhabets.yahtzeescore.data.DataManager
import nl.koenhabets.yahtzeescore.data.Game
import nl.koenhabets.yahtzeescore.databinding.ActivityStatsBinding
import nl.koenhabets.yahtzeescore.view.ScoreView
import org.json.JSONException
import org.json.JSONObject
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.*
import kotlin.math.roundToInt

class StatsActivity : AppCompatActivity() {
    private lateinit var binding: ActivityStatsBinding
    private lateinit var scoreView: ScoreView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityStatsBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        title = getString(R.string.stats)

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        val sharedPref = getSharedPreferences("nl.koenhabets.yahtzeescore", MODE_PRIVATE)
        val game = Game.valueOf(sharedPref.getString("game", "Yahtzee")!!)
        setScoreView(game)
        scoreView.setTotalVisibility(false)
        scoreView.disableEdit()

        var scoreItemsDate = DataManager().loadScores(this, game)

        val entries: MutableList<Entry> = ArrayList()
        val entriesMa: MutableList<Entry> = ArrayList()
        binding.appBarLayout.visibility = View.GONE

        val jsonObject = processScores(scoreItemsDate)
        scoreView.setScores(jsonObject)
        Collections.sort(scoreItemsDate, ScoreComparatorDate())
        scoreItemsDate = scoreItemsDate.reversed()
        var sum = 0f
        val gamesHidden: Int
        if (scoreItemsDate.size > 200) {
            gamesHidden = 30
            binding.textViewStatGraph.text = getString(R.string.games_hidden, gamesHidden)
        } else if (scoreItemsDate.size > 100) {
            gamesHidden = 10
            binding.textViewStatGraph.text = getString(R.string.games_hidden, gamesHidden)
        } else if (scoreItemsDate.size > 50) {
            gamesHidden = 5
            binding.textViewStatGraph.text = getString(R.string.games_hidden, gamesHidden)
        } else {
            gamesHidden = 0
        }
        for (d in scoreItemsDate.indices) {
            Log.i("score", Date(scoreItemsDate[d].date).toString())
            sum += scoreItemsDate[d].score
            val value = sum / (d + 1)
            if (d > gamesHidden - 1) {
                entries.add(Entry(d.toFloat(), value))
            }
        }
        val size = 15
        val movingAverage = MovingAverage(size)
        for (d in scoreItemsDate.indices) {
            movingAverage.addData(scoreItemsDate[d].score.toDouble())
            if (d > 14) {
                entriesMa.add(
                    Entry(
                        d.toFloat(), movingAverage.mean.toFloat()
                    )
                )
            }
        }
        Log.i("entries", entries.size.toString() + "size")
        val dataSet = LineDataSet(entries, getString(R.string.average_score))
        dataSet.color = Color.BLUE
        dataSet.valueTextColor = Color.YELLOW
        val lineData = LineData(dataSet)
        binding.statChart1.data = lineData
        binding.statChart1.description.text =
            getString(R.string.average_score_of_last, scoreItemsDate.size)
        lineChartSetFlags(binding.statChart1)
        val dataSetMa = LineDataSet(entriesMa, getString(R.string.average_score))
        dataSetMa.color = Color.BLUE
        dataSetMa.valueTextColor = Color.YELLOW
        val lineDataMa = LineData(dataSetMa)
        binding.chartMa.data = lineDataMa
        binding.chartMa.description.text = getString(R.string.average_score_of_last_ma)
        lineChartSetFlags(binding.chartMa)
        if (sharedPref.getBoolean("statsInfoDialog", true)) {
            infoDialog(false)
        }
    }

    private fun setScoreView(game: Game) {
        scoreView = ScoreView.getView(game, this)
        binding.activityStatsConstraint.addView(scoreView)
        val set = ConstraintSet()
        set.clone(binding.activityStatsConstraint)
        set.connect(scoreView.id, ConstraintSet.START, ConstraintSet.PARENT_ID, ConstraintSet.START)
        set.connect(scoreView.id, ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP)
        set.connect(binding.statChart1.id, ConstraintSet.TOP, scoreView.id, ConstraintSet.BOTTOM)
        set.applyTo(binding.activityStatsConstraint)
    }

    private fun lineChartSetFlags(lineChart: LineChart) {
        val nightModeFlags = this.resources.configuration.uiMode and
                Configuration.UI_MODE_NIGHT_MASK
        if (nightModeFlags == Configuration.UI_MODE_NIGHT_YES) {
            lineChart.xAxis.textColor = Color.WHITE
            lineChart.axisLeft.textColor = Color.WHITE
            lineChart.legend.textColor = Color.WHITE
            lineChart.description.textColor = Color.WHITE
        }
        lineChart.invalidate()
    }

    private fun infoDialog(checkboxChecked: Boolean) {
        val inflater = this.layoutInflater
        val builder = AlertDialog.Builder(this)
        val view = inflater.inflate(R.layout.dailog_stats, null)
        val checkBox = view.findViewById<CheckBox>(R.id.checkBox)
        checkBox.isChecked = checkboxChecked
        builder.setView(view)
        builder.setPositiveButton("Ok") { _: DialogInterface?, _: Int -> }
        checkBox.setOnClickListener {
            val sharedPref = getSharedPreferences("nl.koenhabets.yahtzeescore", MODE_PRIVATE)
            sharedPref.edit().putBoolean("statsInfoDialog", !checkBox.isChecked).apply()
        }
        builder.show()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.help) {
            val sharedPref = getSharedPreferences("nl.koenhabets.yahtzeescore", MODE_PRIVATE)
            infoDialog(!sharedPref.getBoolean("statsInfoDialog", true))
        } else {
            finish()
        }
        return true
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_stats, menu)
        return true
    }

    private fun processScores(scoreItems: List<ScoreItem>): JSONObject {
        val jsonObject = JSONObject()
        for (k in 1..6) {
            try {
                jsonObject.put(k.toString() + "", processField(scoreItems, k))
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        for (k in 21..29) {
            try {
                jsonObject.put(k.toString() + "", processField(scoreItems, k))
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        return jsonObject
    }

    private fun processField(scoreItems: List<ScoreItem>, d: Int): String {
        var totalScore = 0.0
        var scoreCount = 0.0
        var scoreCountMax = 0.0
        for (it in scoreItems) {
            try {
                val item = it.allScores
                val score = item.getString(d.toString() + "")
                if (score != "" && score != "0") {
                    val valInt = score.toInt()
                    totalScore += valInt
                    scoreCount += 1
                }
                scoreCountMax += 1
            } catch (ignored: JSONException) {
            } catch (ignored: NumberFormatException) {
            }
        }
        if (scoreCount == 0.0) {
            return "0"
        }
        val chance = (scoreCount / scoreCountMax * 100.0).roundToInt()
        val average = totalScore / scoreCountMax
        return round(average, 1).toString() + "(" + chance + ")"
    }

    companion object {
        fun round(value: Double, places: Int): Double {
            var res = 0.0
            if (value != 0.0 && !java.lang.Double.isNaN(value)) {
                require(places >= 0)
                var bd = BigDecimal(value)
                bd = bd.setScale(places, RoundingMode.HALF_UP)
                res = bd.toDouble()
            }
            return res
        }
    }
}